private/buf/bufwire/module_config_reader.go

```diff
@@ -205,6 +216,99 @@ func (m *moduleConfigReader) getModuleModuleConfig(
 	return newModuleConfig(module, config, nil /* Workspaces aren't supported for ModuleRefs */), nil
 }
 
+func (m *moduleConfigReader) getProtoFileModuleSourceConfigs(
+	ctx context.Context,
+	container app.EnvStdinContainer,
+	sourceRef buffetch.SourceRef,
+	workspaceBuilder bufwork.WorkspaceBuilder,
+	configOverride string,
+	externalDirOrFilePaths []string,
+	externalDirOrFilePathsAllowNotExist bool,
+) (_ []ModuleConfig, retErr error) {
+	readBucketCloser, err := m.fetchReader.GetSourceBucket(ctx, container, sourceRef)
+	if err != nil {
+		return nil, err
+	}
+	workspaceConfigs := map[string]struct{}{}
+	for _, workspaceConfigFile := range bufwork.AllConfigFilePaths {
+		workspaceConfigs[workspaceConfigFile] = struct{}{}
+	}
+	moduleConfigs := map[string]struct{}{}
+	for _, moduleConfigFile := range bufconfig.AllConfigFilePaths {
+		moduleConfigs[moduleConfigFile] = struct{}{}
+	}
+	terminateFilePriority := readBucketCloser.TerminateFilePriority()
+	workspaceConfigDirectory := ""
+	moduleConfigDirectory := ""
+	for _, terminateFile := range terminateFilePriority.TerminateFiles() {
+		if terminateFile != nil {
+			if _, ok := workspaceConfigs[terminateFile.ConfigFile]; ok {
+				workspaceConfigDirectory = terminateFile.Path
+				continue
+			}
+			if _, ok := moduleConfigs[terminateFile.ConfigFile]; ok {
+				moduleConfigDirectory = terminateFile.Path
+			}
+		}
+	}
+	// If a workspace and module are both found, then we need to check of the module is within
+	// the workspace. If it is, we use the workspace. Otherwise, we use the module.
+	moduleInWorkspace := false
+	if workspaceConfigDirectory != "" && moduleConfigDirectory != "" {
+		workspaceConfig, err := bufwork.GetConfigForBucket(ctx, readBucketCloser, readBucketCloser.RelativeRootPath())
+		if err != nil {
+			return nil, err
+		}
+		for _, dir := range workspaceConfig.Directories {
+			workspaceDirectory := filepath.Join(workspaceConfigDirectory, dir)
+			relativePath, err := filepath.Rel(workspaceDirectory, moduleConfigDirectory)
```

Comment:
Had an offline conversation with @amckinney , going to clean up the use of `normalpath` in the PR.